modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -1,64 +1,155 @@
+            .as(none)
+        }
+
+      def pullGlobalSnapshots: F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        maybeMajorityPeerIds
+          .fold(pullGlobalSnapshotsFromRandomPeer)(pullGlobalSnapshotsFromMajority)
+          .handleErrorWith { e =>
+            logger
+              .warn(e)(s"Failure pulling global snapshots")
+              .as(
+                List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[LatestSnapshotTuple]
+              )
+          }
+
+      private def pullGlobalSnapshotsFromMajority(
+        majorityPeerIds: NonEmptyList[PeerId]
+      ): F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        lastGlobalSnapshotStorage.getOrdinal.flatMap {
+          _.fold {
+            pullLatestSnapshotWithMajorityHash(majorityPeerIds).map(_.asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+          } { lastStoredOrdinal =>
+            for {
+              msd <- getLatestMajoritySnapshotData(majorityPeerIds)
+              l0Peers <- globalL0ClusterStorage.getPeers
+              pulled <- pullVerifiedSnapshots(lastStoredOrdinal.next, msd, l0Peers)
+            } yield pulled.asRight[LatestSnapshotTuple]
+          }
+        }
+
+      private def pullLatestSnapshotWithMajorityHash(
+        majorityPeerIds: NonEmptyList[PeerId]
+      ): F[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)] = {
+        type Agg = SortedSet[L0Peer]
+        type Result = LatestSnapshotTuple
+        for {
+          peers <- globalL0ClusterStorage.getPeers.map(nes => Random.shuffle(nes.toSortedSet))
+          majorityPeers <- getL0Peers(majorityPeerIds)
+          result <- peers.tailRecM[F, Result] { l0Peers =>
+            l0Peers.headOption.fold {
+              NoPeersWithMajorityHash.raiseError[F, Either[Agg, Result]]
+            } { l0Peer =>
+              pullLatestSnapshotFromPeer(l0Peer).flatMap { latest =>
+                Applicative[F].ifF(verifyLatestSnapshot(latest, majorityPeers))(
+                  latest.asRight[Agg],
+                  l0Peers.tail.asLeft[Result]
+                )
+              }.handleErrorWith { err =>
+                logger
+                  .warn(err)(s"Error pulling latest snapshot from peer ${l0Peer.show}")
+                  .as(l0Peers.tail.asLeft[Result])
+              }
+            }
+          }
+        } yield result
+      }
+
+      private def verifyLatestSnapshot(
+        snapshot: LatestSnapshotTuple,
+        majorityPeers: NonEmptyList[L0Peer]
+      ): F[Boolean] =
+        List(
+          StateProofValidator.validate(snapshot._1, snapshot._2).map(_.isValid),
+          getMajorityOrdinal(majorityPeers).map(_.fold(false)(ord => ordinalRange.contains(ord.value - snapshot._1.ordinal.value))),
+          getMajorityHash(majorityPeers, snapshot._1.ordinal).map(_.contains(snapshot._1.hash))
+        ).reduce(_.ifM(_, false.pure[F]))
```

Comment:
maybe:
```
        (
          StateProofValidator.validate(snapshot._1, snapshot._2).map(_.isValid),
          getMajorityOrdinal(majorityPeers).map(_.fold(false)(ord => ordinalRange.contains(ord.value - snapshot._1.ordinal.value))),
          getMajorityHash(majorityPeers, snapshot._1.ordinal).map(_.contains(snapshot._1.hash))
        ).mapN(_ && _ && _)
```

Reply:
I'm not sure if your implementation will short circuit.

Reply:
@TheMMaciek I have just checked and David's solution will short circuit.
As alternative you could play with `existsM` and negation:

`.existsM(_.map(!_))`

Reply:
@TheMMaciek I tested it yesterday. Here's a couple of tests that show the `reduce` implementation stops printing on false whereas the `mapN` does not.
```
  import cats.syntax.all._
  def foo(n: Int): IO[Boolean] =
    IO(println(n)).as(n % 2 == 0)

  test("reduce") { _ =>
    List(foo(0), foo(1), foo(2))
      .reduce(_.ifM(_, false.pure[IO]))
      .as(success)
  }

  test("mapN") { _ =>
    (foo(4), foo(5), foo(6))
      .mapN(_ && _ && _)
      .as(success)
  }
```

Reply:
Cool, approved. You can consider `existsM(_.map(!_))` to get rid of explicit `false` fallback but up to you.

Reply:
Right, I was aware that mapN won't short circuit, wasn't aware that reduce will, cool!